### ISSUE TITLE
Binary Crossentropy Loss is only for binary classes

### ISSUE THREAD

**Issue by yCobanoglu** [pnyaio/VQVAE_Flax#1]
Hi,
thanks for the tutorial. I was working through it and saw `loss = optax.sigmoid_binary_cross_entropy(logits, batch[0]).mean()` for the vanilla autoencoder. I think one should use l2 loss here cause BinaryCrossentropy is only for labels with 0-1.

**Comment by yCobanoglu** [comment]
I just realized that MNIST is binary not greyscale...

**Comment by aillaud** [comment]
Hello,
I'm glad my experiments can somewhat be useful for someone else 
Regarding your question, it is true that the BCE is the perfect function for binary classification, and that MSE is usually preferred for reconstruction tasks. However, BCE is a perfectly valid loss function for autoencoders and is used in many tutorials on MNIST : https://github.com/pytorch/examples/blob/main/vae/main.py, https://blog.keras.io/building-autoencoders-in-keras.html, ...
The choice of loss function actually reflects your assumption about the distribution of the predictions of the model (BCE for multinomiale, MSE for Gaussian, MAE for Laplacian). In practice, BCE works for autoencoders on MNIST.
Some links with more detailled explanations if you want to dig further:

- https://stats.stackexchange.com/questions/370179/why-binary-crossentropy-can-be-used-as-the-loss-function-in-autoencoders
- https://stackoverflow.com/questions/65223762/variational-autoencoders-mse-vs-bce
- https://stats.stackexchange.com/questions/245448/loss-function-for-autoencoders
